feat(js): Add micro-frontend docs (#6754)
getsentry/sentry-docs: src/platforms/javascript/common/configuration/micro-frontend-support.mdx

---
title: Micro Frontend Support
sidebar_order: 200
description: ''
keywords: ['micro frontend', 'multiplexed transport']
---

The recommended way to use the Sentry SDK with Micro Frontend is to set up a multiplexed transport when calling `Sentry.init`. The multiplexed transport will send events to different Sentry projects based on the attributes on that event, which is how you can isolate events from a specific micro-frontend to a specific Sentry project.

> In the example below, we use the `feature` tag to determine which Sentry project to send the event to by using different DSNs. If the event does not have a `feature` tag, we send it to the fallback DSN defined in `Sentry.init`.

Reviewer: ```suggestion
The example below uses the `feature` tag to determine which Sentry project to send the event to by using different DSNs. If the event does not have a `feature` tag, we send it to the fallback DSN defined in `Sentry.init`.
```

> ```ts

Reviewer: ```suggestion
```typescript
```

import { makeMultiplexedTransport } from '@sentry/core';
import { init, captureException, makeFetchTransport } from '@sentry/browser';

function dsnFromFeature({ getEvent }) {
  const event = getEvent();
  switch(event?.tags?.feature) {
    case 'cart':
      return ['__CART_DSN__'];
    case 'gallery':
      return ['__GALLERY_DSN__'];
  }
  return []
}

init({
  dsn: '__FALLBACK_DSN__',

>   transport: makeMultiplexedTransport(makeFetchTransport, dsnFromFeature)

Reviewer 2: Random question but might be relevant to document: which ones are called first: the "matcher" ("dsnFromFeature") or the init's "beforeSend"?

As an example, if I enhance my event tags at init's `beforeSend` (e.g. I add "event.tags.feature = cart"), will the enhanced event tags be available at this matcher (i.e. at `dsnFromFeature` when `getEvent()` is called)?

Context: before this multiplexed transport feature is available, one workaround to send different event to different Sentry project is via `beforeSend` and initialize a BrowserClient there (https://github.com/sentry-demos/sentry-micro-frontend/blob/5996c4b528f58696b729db9d41f4962a83c53c77/methods/simple-remote.js#LL42C28-L42C28) So when implementing this mechanism at my company, my implementation actually adds more tags there than just finding out relevant DSN to send to (e.g. I added the micro app's release info). If I could access the tags added in "beforeSend" when the matcher is called, I could reduce some code duplication. However if that's not possible, it's also ok too for me.

Thanks~

Author: The init's `beforeSend` is called before the matcher's `dsnFromFeature`. Will call this out in the docs - thanks for updating it!

});

captureException(new Error('oh no!', scope => {
  scope.addTag('feature', 'cart');
  return scope;
});
```

You can then set tags/contexts on events in individual micro-frontends to decide which Sentry project to send the event to.

<Alert level="warning" title="Note">

> Currently there is no way to isolate tags/contexts/breadcrumbs/spans to events sent to a specific Sentry project. This is because browsers do not have support for async contexts, which is required for this functionality to work. Once the [tc39 Async Context](https://github.com/tc39/proposal-async-context) proposal is implemented in all browsers, we will be able to add this functionality.

Reviewer: ```suggestion
Currently, there is no way to isolate tags/contexts/breadcrumbs/spans to events sent to a specific Sentry project. This is because browsers do not have support for async contexts, which is required for this functionality to work. Once the [tc39 Async Context](https://github.com/tc39/proposal-async-context) proposal is implemented in all browsers, we will be able to add this functionality.
```

</Alert>

> ### makeMultiplexedTransport API

Reviewer: ```suggestion
## `makeMultiplexedTransport` API
```

Reviewer: bringing all subheadings out by a level, so h2 is used

Reviewer: also, I think code formatting works in headers.

`makeMultiplexedTransport` takes an instance of a transport (we recommend `makeFetchTransport`) and a matcher function that returns an array of DSNs.

```typescript
type EnvelopeItemType =
  | 'client_report'
  | 'user_report'
  | 'session'
  | 'sessions'
  | 'transaction'
  | 'attachment'
  | 'event'
  | 'profile'
  | 'replay_event'
  | 'replay_recording'
  | 'check_in';

interface MatchParam {
  /** The envelope to be sent */
  envelope: Envelope;
  /**
   * A function that returns an event from the envelope if one exists. You can optionally pass an array of envelope item
   * types to filter by - only envelopes matching the given types will be multiplexed.
   *
   * @param types Defaults to ['event'] (only error events will be matched)
   */
  getEvent(types?: EnvelopeItemType[]): Event | undefined;
}

type Matcher = (param: MatchParam) => string[];

declare function makeMultiplexedTransport(
  transport: (options: TransportOptions) => Transport,
  matcher: Matcher
): (options: TransportOptions) => Transport;
```

> ### Examples

Reviewer: ```suggestion
## Examples
```

> #### Allowing individual Micro Frontends to add their own DSNs

Reviewer: ```suggestion
### Allowing Individual Micro Frontends to Add Their Own DSNs
```

Reviewer: Use title casing on headers

> In the example below, we have a list of DSNs and tags. We use the `tags` attribute on the event to determine which Sentry project to send the event to by using different DSNs. If the event does not have a `tags` attribute, we send it to the fallback DSN defined in `Sentry.init`. Individual micro-frontends can then mutate the `myDsns` variable to add their own DSNs and tags.

Reviewer: ```suggestion
The example below gives a list of DSNs and tags. It uses the `tags` attribute on the event to determine which Sentry project to send the event to by using different DSNs. If the event does not have a `tags` attribute, we send it to the fallback DSN defined in `Sentry.init`. Individual micro-frontends can then mutate the `myDsns` variable to add their own DSNs and tags.
```

> ```ts

Reviewer: ```suggestion
```typescript
```

const myDsns = [
  { dsn: 'https://...', tag: 'my-tag' },
  { dsn: 'https://...', tag: 'other-tag' },
];

function dsnFromTag({ getEvent }) {
  const event = getEvent();
  const opt = myDsns.find(({tag} => event?.tags?.includes(tag));
  return opt ? [opt.dsn] : [];
}

init({
  dsn: '__FALLBACK_DSN__',
  transport: makeMultiplexedTransport(makeFetchTransport, dsnFromTag)
});
```

> #### Sending the same event to multiple DSNs

Reviewer: ```suggestion
### Sending the Same Event to Multiple DSNs
```

> Since the matcher returns an array, you can enforce that an event is sent to

Reviewer: ```suggestion
Since the matcher returns an array, you can enforce that an event is sent to multiple DSNs.
```

Reviewer: I'm assuming this was what the rest of the sentence was meant to say?

Author: yup good catch

> ```ts

Reviewer: ```suggestion
```typescript
```
